Add post "Calculating type sets is harder than you think" (#3)
Merovius/merovius.github.io: content/posts/2022-05-13-calculating-type-sets.markdown

all time to the language: [Generics](https://go.dev/doc/go1.18#generics). If
you want a comprehensive introduction to the topic, there are many out there
and I would personally [recommend this talk I gave at the Frankfurt Gopher
Meetup](https://www.youtube.com/watch?v=QP6v-Q5Foek).

This blog post is not an introduction to generics, though. It is about [this
sentence from the spec](https://go.dev/ref/spec#Operands):

> Implementation restriction: A compiler need not report an error if an
> operand's type is a type parameter with an empty type set.

This restriction might seem strange to you. After all, if a type set is empty,
it seems very helpful to report that to the user, because they obviously made a
mistake - an empty type set can never be used as a constraint, as the function
using it could never be instantiated.

I want to explain the reason that sentence is there and also go into a couple
of related design decisions of the generics design. I'm trying to be expansive
in my explanation, which means that you should not *need* any special knowledge
to understand it. It also means, some of the information might be boring to
you - feel free to skip the corresponding sections.

> The reason that sentence is in the Go spec, is that it turns out to be hard to

Reviewer: The comma trips me up. Maybe start like "This sentence is in the Go spec because"...?

actually determine if a type set is empty. Hard enough, that the Go team did
not want to require an implementation to solve that. Let's see why.

## P vs. NP

When we talk about whether or not a problem is hard, we often group problems
into two big classes:

1. Problems which can be *solved* reasonably efficiently. This class is called
   P.
2. Problems which can be *decided* reasonably efficiently. This class is called
   NP.

The first obvious follow up question is “what does ‘reasonably efficient’
mean?”. The answer to that is “there is an algorithm with a running time
polynomial in its input size”[^1].

The second obvious follow up question is “what's the difference between
‘solving’ and ‘deciding’?”.

*Solving* a problem means what you think it means: Finding a solution. If I
give you a number and ask you to solve the factorization problem, I'm asking
you to find a (non-trivial) factor of that number.

> *Deciding* a problem means that I give you solution and I'm asking you if the

Reviewer 2: ```suggestion
*Deciding* a problem means that I give you a solution and I'm asking you if the
```

solution is correct. For the factorization problem, I'd give you two numbers
and ask you to verify that the second is a factor of the first.

These two things are often very different in difficulty. If I ask you to give
me a factor of 297863737, you probably know no better way than to sit down and
try to divide it by a lot of numbers and see if they divide it evenly. But if I
ask you to verify that 9883 is a factor of that number, you just have to do a

> bit of long division and it either divides it, or it does not.

Reviewer 3: nice explanation of P vs NP. I wish they had done it this way at uni.

Author: Thanks :) TBF, it's just an informal explanation, so I'm not bogged down by formalism. Then again, nothing would prevent professors from giving *both* an informal explanation *and* a formal proof - and the ones I always liked best tended to do that.

It turns out, that every problem which is efficiently *solvable* is also
efficiently *decidable*. You can just calculate the solution and compare it to
the given one, if you want to decide it. So, every problem in P is also in
NP[^2]. But it is [a famously open question](https://en.wikipedia.org/wiki/P_versus_NP_problem)
whether the opposite is true - that is, we don't really *know*, if there are
problems which are hard to solve, but easy to decide.

The reason this is hard to know in general, is that just because *we have not
found an efficient algorithm to solve a problem*, does not mean there is none.
In practice, though, we usually assume that there are some problems like that.

One fact that helps us talk about hard problems, is that it turns out that
there are some problems which are *as hard as possible* in NP. That means we

> were able to prove, that if you can solve one of these problems, you can use

Reviewer 2: ```suggestion
were able to prove that if you can solve one of these problems you can use
```

> that to solve *any other problem in NP*. These problems are called “NP-complete”.

Reviewer 4: This and the previous paragraphs explained P vs NP to me in a way I'd never read before, very well done!

Author: Thanks :)

That is, to be frank, plain magic and explaining it is far beyond my
capabilities. But it helps us to tell if a given problem is hard, by doing it
the other way around. If solving problem X would enable us to solve one of

> these NP-complete problems, than solving problem X is obviously itself NP-complete

Reviewer 2: ```suggestion
these NP-complete problems then solving problem X is obviously itself NP-complete
```

> these NP-complete problems, than solving problem X is obviously itself NP-complete

Reviewer 3: ```suggestion
these NP-complete problems, then solving problem X is obviously itself NP-complete
```

and therefore *probably very hard*. This is called a “proof by reduction”.

One such problem, which is very often used to prove a problem is hard, is
boolean satisfiability.

## SAT

Imagine I give you a formula. This formula has a bunch of variables, which can
all be `true` or `false`, joined by logical operators. For example

    ((¬X∧Y)∨Z)∧(X∨¬Y)

Or, if you are more familiar with programming syntax:

    ((!x && y) || z) && (x || !y)

> If I give you an assignment to these variable, you can reasonably efficiently

Reviewer 2: ```suggestion
If I give you an assignment to these variables, you can efficiently
```

I think we could call it "efficient" by almost any metric :)

tell me, if the formula evaluates to `true` or `false`, by substituting them in
and evaluating the formula. For example

    X = true
    Y = false
    Z = true
    ((¬X∧Y)∨Z)∧(X∨¬Y) = ((¬true ∧ false) ∨ true)∧(true ∨ ¬false)
                      = ((false ∧ false) ∨ true)∧(true ∨ true)
                      = (false ∨ true) ∧ true
                      = true ∧ true
                      = true

But if I *only* give you the formula and ask you to *find* an assignment which

> makes it `true` - or if such an assignment exist - you probably have to try out

Reviewer 2: ```suggestion
makes it `true` - or even to find whether such an assignment exists - you probably have to try out
```

all possible assignments to see if any of them does. That's easy for three
variables, but there are 2<sup>n</sup> possible assignments, so it takes
*exponential* time in the number of variables.

This problem is called "boolean satisfiability" and it is NP-complete.

It turns out to be extremely important *in what form* the formula is given,
though. Some forms make it pretty easy to solve, while others make it hard.

> For example, every formula has what is called a [“Disjunctive Normal Form”

Reviewer 2: The word "has" is pulling more weight than I expected here.
Maybe make it a bit clearer with something like:

```suggestion
For example, every formula has (can be rewritten into) what is called a [“Disjunctive Normal Form”
```

(DNF)](https://en.wikipedia.org/wiki/Disjunctive_normal_form), so called
because it consists of a series of conjunction (“and”) terms, joined together
by disjunction (“or”) operators[^3]:

    (X ∧ Z) ∨ (¬Y ∧ Z)

Each term has a subset of the variables, possibly negated, joined by ∧. The
terms are then joined together using `∨`.

> Solving the satisfiability problem for a formula in DNF is obviously very easy:

Reviewer 2: ```suggestion
Solving the satisfiability problem for a formula in DNF is easy:
```

1. Go through the individual terms. `∨` is true if and only if either of its
   operands is true, so for each term:
   * If it contains both a variable and its negation (i.e. `X∧¬X`) it can
     never be true, continue to the next term.

>    * Otherwise, you can read an assignment from the term - assign `true`, if

Reviewer 2: ```suggestion
   * Otherwise, you can infer an assignment from the term: assign `true` if
```

There's a bit of an ambiguity going on here. In some places, "an assignment" implies an assigment to all the variables (e.g. "an assignment to these variables" on line 105), but here it's an assignment to one of them.

For consistency with programming languages, I think I'd go with an assignment meaning an assignment to a single variable, so line 105 would read "an assignment to each of these variables".

Given the above, maybe something like this could work?

```
   * Otherwise you can read two assignments from the term: for each variable in the term, if it's of the form `X` assign `true` to `X`, otherwise assign `true` to `X`.
```

Author: I re-wrote this. I went with `assignment` always referring to a simultaneous assignment to all variables, to stay consistent with the rest of the article (I hope).
(I'll push it in a bit. Sure wish github reviews would support draft comments)

     the term contains `X` and assign `false`, if it contains `¬X`. The term
     is then true, so the entire formula is.
2. If none of the terms can be made `true`, the formula can't be made `true`
   and there is no solution.

On the other hand, there is also a [“Conjunctive Normal
Form”](https://en.wikipedia.org/wiki/Conjunctive_normal_form). Here, the
formula is a series of *disjunction* terms, joined together with a
*conjunction*:

    (¬X ∨ Z) ∧ (Y ∨ Z) ∧ (X ∨ ¬Y)

For this, the same idea does not work - you have to take *all terms* into
account simultaneously, to find an assignment and can't just tackle them one by
one. In fact, solving satisfiability on CNF (often abbreviated as “CNFSAT”) is
NP-complete[^4].

Very often, when we want to prove a problem is hard, we do so by reducing it to
CNFSAT. That's what we will do for the problem of calculating type sets. But
there is one more preamble we need.

## Sets and Satisfiability

There is an important relationship between
[sets](https://en.wikipedia.org/wiki/Set_(mathematics)) and boolean functions
(functions which return `true` or `false`).

Say we have some universe of objects `U`. And say we have a boolean function

> `f:U→B` with `B:={true,false}`. We can then create a set from that, by looking

Reviewer 2: Maybe worth explaining the notation here or just using Go notation given that most people reading this will be familiar with go? (`func f(U) bool`) ?

Author: I admit that I assume a bit of mathematical notation in this article where I probably should not. The issue I'm having is that many of these concepts don't have a clear analog in Go. So using Go syntax consistently is out. It seems better to then consistently use "math syntax", when I talk about formal concepts and Go syntax when talking about Go.

However, if readers are not familiar with the notation, *something* clearly needs to happen. Perhaps a comment every time I introduce a piece of notation (but I'm worried that this blows up the article even further and hinders the reading flow).

I'll leave this open as a reminder, for now, while I ponder it a bit.

Reviewer 4: I agree with Roger FWIW

Reviewer: I also agree - but this will be hard for the set definition `S[f]` below; a for loop will not cut it.
Still, if you stay with the math notation, you might consider changing "boolean functions" to "predicates" - preserving the explanation. Reading "boolean function" and "return `true` or `false`" somehow triggers programming mode for me.

at all objects for which `f` is `true`:

    S[f] := { x∈U | f(x) }

> We use the index form `S[f]` to indicate, that `S` depends on the function `f`.

Reviewer 2: ```suggestion
We use the index form `S[f]` to indicate that `S` depends on the function `f`.
```

> Likewise, if we have any set `S⊆U`, we can create a boolean function from it:

Reviewer 2: Again, maybe explain the subset notation for people that might not be familiar with it?

    f[S]: U → B

>           x ↦ x∈S

Reviewer 2: I find this a bit confusing. I presume that the indented line is the body of the function, and that the `↦` symbol has arguments on the left and an expression to evaluate on the right, but this isn't necessarily clear to those without a maths background.

Reviewer 5: Yeah, I'm with Roger, it's unclear to me what this means.

Reviewer: I also didn't know that notation (or forgot it). Same for the index form.

That is, the function returns `true` if an element is in `S` and `false`
otherwise.

An interesting aspect of this, is that we can translate between boolean
operators and set operators this way. For example, set union `∪` becomes
equivalent to logical disjunction `∨`:

>     S[f] ∪ S[g] = { x∈U | f(x) } ∪ { x∈U | g(x) }

Reviewer 2: Likewise, it would be good to explain the `|` notation. Casting back to my own long distant background, isn't there a missing `∀` symbol before the `x∈U` expressions?

Author: Notations vary, some people use `{ x∈U: f(x) }` instead, for example. But I think you mix this up with `∀x∈U: f(x)`, which means something different - I don't think ∀ is commonly used with set-notation.

Either way, as above, clearly something needs to be fixed here.

Reviewer 4: Around about here I'm lost. What is the target audience of this post?

Author: @johanbrandhorst Well, you, among others. I do genuinely want this to be as broadly accessible as possible. So, yes. As the notation stands in the way of that, I'll try and figure something out.

                = { x∈U | f(x) ∨ g(x) }

    f[S](x) ∨ f[T](x) = x∈S ∨ x∈T = x∈(S∪T)

> Similarly, `∩` becomes equivalent to `∧`, `⊆` becomes equivalent to `⇒` and `¬`

Reviewer 2: ```suggestion
Similarly, `∩` becomes equivalent to `∧`, `⊆` becomes equivalent to `⇒`, and `¬`
```

The "`⊆` becomes equivalent to `⇒`" part isn't clear to me.

Author: Yeah, thought that might be confusing. I just cut it.

Arguably, it's just the definition of `⊆`. `A⊆B` is defined as `a∈A ⇒ a∈B` - "if something is in `A`, it must also be in `B`. That's part of why I don't know how I'd explain the `⊆` notation - in a way, explaining the notation is the point of this section.

But yeah. Anyways, for now, I just cut this :)

> becomes equivalent to set complement `U\S := {x∈U | x∉S}`.

Reviewer 2: where `\` is the set complement operator?
Why are we using `:=` here when we've used `=` up to now?

Author: Bad habits die hard. I'm used to just throw in a definition, if I haven't used an operator before and have people know what I mean. That is, the `:=…` part is "explaining what `\` is" :)

I had some trouble when writing this, because there is no well-known single symbol for "set complement", I think.

Changed this, let me know if you have any better ideas (once I pushed).

The boolean formulas we looked at in the previous section are a form of these
boolean functions. They take some boolean arguments and return a boolean, e.g.

>       f: B³ → B

Reviewer 2: `B³` is new notation here. Maybe avoid if possible?

Author: Any suggestion? `B×B×B` didn't seem any more useful.

Reviewer 5: f(B<sub>1</sub>, B<sub>2</sub>, B<sub>3</sub>) ?

>     (X,Y,Z) ↦ ((¬X∧Y)∨Z)∧(X∨¬Y)

Reviewer 2: should this be indented for consistency with the previous instance of this kind of notation?

Author: The usual convention is to align the arrows. Maybe I should just commit and use MathJax :)

> It turns out that *every* boolean function with boolean parameters can be

Reviewer 2: It would be quite nice if the "It turns out" was a link to something that goes further into this result.

written as a logical formula - in particular, every such function has a CNF and
a DNF.

If we want to do that for other sets (apart from `B^n`, that is), we first have
to be able to translate it into a set of boolean variables. As we will see,
this can be easily done in the case we are interested in.

All of this allows us to talk about what we mean by “computing a set is hard”.
With this equivalency, checking if a given object is in a set becomes the

> *decision* problem of SAT - `x∈S`, exactly if `f[S](x)` evaluates to `true`.

Reviewer 2: I'm not sure exactly what this is saying, particularly the "exactly if" is confusing to me. Perhaps be a bit more verbose here and include the word "decide" in the explanation?

And checking if a set is empty becomes the *solution* problem of SAT - `S=∅`,
exactly if `f[S]` is not satisfiable.

With this, let us look at the specific sets we are interested in.

## Basic interfaces as type sets

Interfaces in Go have always described sets of types. For example, the interface

```go
type S interface {
    X()
    Y()
    Z()
}
```

is “the set of all types which have a method `X()` and a method `Y()` and a method `Z()`”.

> We also had a way to express set intersection, using [interface embedding](https://go.dev/ref/spec#Embedded_interfaces):

Reviewer 2: I think it's maybe better to use the present tense, as this is all still very much true today.

```suggestion
We also have a way to express set intersection, using [interface embedding](https://go.dev/ref/spec#Embedded_interfaces):
```

```go
type S interface { X() }
type T interface { Y() }
type U interface {
    S
    T
}
```

This expresses `T=S∩T` as an interface. We can also view the predicate “has a
method `X()`” as a boolean variable and think of this as the formula `X∧Y`.

Surprisingly, there is also a simple form of negation. It happens implicitly,
because a type can not have two methods with the same name, but different
types. So, implicitly, if a type has a method `X()`, it does *not* have a
method `X() int`, for example[^5]:

```go
type X interface { X() }
type NotX interface{ X() int }
```

> This meant that you could express interfaces which could never be satisfied by

Reviewer 2: ```suggestion
This meant that even without type parameters, we can express interfaces which can never be satisfied by
```

Author: I don't think type parameters are relevant.

> any type (i.e. which described an empty type set):

Reviewer 2: ```suggestion
any type (i.e. which describe an empty type set):
```

```go
interface{ X; NotX }
```

[The compiler rejects such interfaces](https://go.dev/play/p/r4kpXNynscX).

> So, we pretty much had a language to build boolean formulas, where every

Reviewer 2: ```suggestion
So, we pretty much had a language to build boolean formulas, where every
```

I'm not keen on the "pretty much" here. Do we or don't we?

Author: Well, we "pretty much" have :) It's not really a language which expresses formulas, it is a language which expresses sets. But it also doesn't *really* expresses formulas, it only "pretty much" expresses them (e.g. we don't have Tertium Non Datur).

Anyways, you're right.

formula had the form

    X∧Y∧¬Z

> Checking if a type implements an interface meant to check if it is in the set

Reviewer 2: ```suggestion
Checking if a type implements an interface checks if it is in the set
```

Author: Changed this to present tense, but I'm happier with the "means"

described by that interfaces. Which meant checking if the corresponding formula
is satisfied.

And checking if an interface is invalid meant checking if the set it describes
is empty. Which meant checking if the formula is *satisfiable*. This was easy,

> though. If you look closely, those formulas are in DNF and as we noted above,

Reviewer 2: "If you look closely" - I think it would be good to point out the equivalence explicitly rather than leaving it as an exercise for the reader.

solving SAT on a formula given in DNF is easy.

## Adding unions

Go 1.18 extends the interface syntax. For our purposes, the important addition
is the `|` operator:

```go
type S interface{
    A | B
}
```

This represents the set of all types which are in the *union* of the type sets
`A` and `B` - that is, the set of all types which are in `A` *or* in `B` (or
both).

> This means our language of expressible formulas now also includes a

Reviewer 2: I think this could expanding: why does `A | B` imply the `∨` operator? I think you need to say explicitly that `A | B` allows all operations provided by both `A` and `B`.

Author: Added an explanation. It's the opposite, BTW. "Allowed operations" are dual to "type sets", so the union of type sets allows all operations which are in *both* `A` and `B`.

`∨`-operator. What's more, the form of our formula is now a *Conjunctive*
normal form - every line is a term of `∨` and the lines are connected by `∧`:

```go
type X interface { X() }
type NotX interface{ X() int }
type Y interface { Y() }
type NotY interface{ Y() int }
type Z interface { Z() }
type NotZ interface{ Z() int }

// (¬X ∨ Z) ∧ (Y ∨ Z) ∧ (X ∨ ¬Y)

> type S interface {

Reviewer 2: I think you need to say somewhere that this is actually prohibited by the [current spec](https://go.dev/ref/spec#Interface_types):

> Implementation restriction: A union (with more than one term) cannot contain the [predeclared identifier](https://go.dev/ref/spec#Predeclared_identifiers) comparable or interfaces that specify methods, or embed comparable or interfaces that specify methods
 
This is kinda important, as it undermines the whole premise of this article - it means that calculating a type set is Go _isn't_ actually harder than one might think, with the current spec at least.

Author: I understand the argument, but to me, this is an important structural decision for the flow of the article. I'm trying to follow the reasoning behind the evolution of the design and I'm doing that piece by piece, not assuming things I haven't introduced. Talking about this statement and why it (hopefully) solves the problems requires already having talked about the union operator and embedding and it also requires knowing the problem.

What's more, I don't fully agree with the statement that calculating a type set is not harder than one might think. I believe it is still harder than one might think, as there are still subtleties involved and everything I mention still needs to be kept in mind. And while it *might* not be NP-complete anymore, we don't actually know that. At least as far as I know.

FWIW, I originally mentioned this in the introduction, but I couldn't get it to a point where I was happy with it. Once I removed it and decided to move it to its own section in the back, the overall structure I wanted became far more clear.

I see the criticism and I'm trying to acknowledging it by calling what I'm doing a bait-and-switch. But I'm claiming creative freedom here :)

    NotX | Z
    Y | Z
    X | NotY
}
```

There is one detail which becomes important here. There are types which
*neither* implement `X` *nor* do they implement `NotX` - if they don't have an
`X` method at all, for example. If we want to really prove that calculating

> these new type sets is NP complete, we need to fix that. We want that a type

Reviewer 2: s/want/need/ ?

Also, why exactly _is_ this a problem for determining the type set?

Author: I'm not sure I understand the question.

Throughout the article, I kind of talk about "calculating the type set" and "checking if a type set is empty" as sort of the same thing. Is that what you are asking? Because that does probably need clarification (I'm not sure where and how).

set is empty *exactly* if its formula is not satisfiable. Boolean variables are
always either `true` or `false`, though.

> “Luckily”, our new operator gives us a way to fix that:

Reviewer 2: Which new operator? `|` ? Maybe explain something about tertium non datur and how it's relevant here (I had to look up the term) ?

```go
type TertiumNonDatur interface {
    X | NotX
    Y | NotY
    Z | NotZ
}

// (¬X ∨ Z) ∧ (Y ∨ Z) ∧ (X ∨ ¬Y)
type S interface {
    TertiumNonDatur

    NotX | Z
    Y | Z
    X | NotY
}
```

> Now, any type we consider necessarily has either an `X()` or an `X() int`

Reviewer 2: Coming at this from fresh, this is hard to credit, because there are clearly many types that don't have an `X` method at all. Perhaps you need to make it clear that we're only looking at a limited set of concrete types or something?

Author: That's what `TertiumNonDatur` does - it guarantees that anything that could possibly implement `S` must have an `X` method returning either nothing, or an `int`.

So, hopefully this will be clear once I explain TND a bit more.

method, so either implements `X`, or implements `NotX` (and never both).

> So, checking if these general type sets are empty is NP complete.

Reviewer 2: Make this more explicit?
```suggestion
So because checking that an interface has an empty type set is equivalent to checking that a boolean formula in CNF has no solutions, which is NP complete as described above.
```

Even worse, as the definition of what operations you can do on a variable of
type parameter type, [we can show that *even type-checking a generic function*
is NP-complete](https://github.com/golang/go/issues/45346#issuecomment-822330394)
with this language.

## Why do we care?

A fair question is “why do we even care? Surely these cases are super exotic
and in any real program, checking this is trivial”.

That's true, but there are still reasons to care:

- We still have to at least *talk about* type sets, because they also define
  the set of valid operations on a type parameter constrained by it. We still
  have to be sure that we can do all the necessary computations in the type
  checker.
- Go has the goal of having a fast compiler. And importantly, one which is
  guaranteed to be fast *for any program*. If I give you a Go program, you can
  be reasonably sure that it compiles quickly, in a time frame predictable by
  the size of the input.

  If I *can* craft a program which compiles slowly, this is no longer true.

  This is especially important for environments like the Go playground, which
  regularly compiles untrusted code.

> - NP complete problems are notoriously hard to debug, if they fail.

Reviewer 2: ```suggestion
- NP complete problems are notoriously hard to debug if they fail.
```

  If you use Linux, you might have occasionally run into a problem where you
  accidentally tried installing conflicting versions of some package. And if
  so, you might have noticed that your computer first chugged along for a while
  and then gave you an unhelpful error message about the conflict. And maybe
  you had trouble figuring out which packages declared the conflicting
  dependencies.

>   This is something vrey typical for NP complete problems. As an exact solution

Reviewer 3: ```suggestion
  This is something very typical for NP complete problems. As an exact solution
```

>   This is something vrey typical for NP complete problems. As an exact solution

Reviewer 2: ```suggestion
  This is typical for NP complete problems. As an exact solution
```

  is often too hard to compute, they rely on heuristics and randomization and
  it's hard to work backwards from a failure.

> - It would require to somehow specify how a compiler should implement this

Reviewer 2: This doesn't read great. What exactly would require the implementation specification? In general, the spec doesn't mention implementation algorithms, right?

Author: Rewrote this to make it clearer. In general, the spec does not *need* to mention an implementation algorithm, because there are correct and efficient choices. If there is no correct and efficient choice, we get the dichotomy "we have to prescribe a heuristic" or "programs might stop compiling when the heuristic changes or differs between compilers".

  check.
  As an exact solution is not possible, there needs to be some heuristic (even
  if it is just “give up after a bit”). Which either needs to be put into the
  language specification, which is very complex, for little benefit. Or
  we leave it up to the implementation, in which case a program might stop
  compiling when changing Go compilers.

## The fix

Given that there must not be an NP-complete problem in the language
specification and given that Go 1.18 was released, this problem must have
somehow been solved.

What changed is that the language for describing interfaces was limited from
what I described above. [Specifically](https://go.dev/ref/spec#General_interfaces)

> > Implementation restriction: A union (with more than one term) cannot contain

Reviewer 2: I think this should be mentioned earlier, as I mentioned above. Best to be up front that this is only a theoretical problem currently.

  one. Thus, being able to decide a problem on a normal computer, means being
  able to solve it on a non-deterministic one.

[^3]: You might complain that it is hard to remember if the “disjunctive normal
  form” is a disjunction of conjunctions, or a conjunction of disjunctions -
  and that no one can remember which of these means “and” and which means “or”
  anyways. You would be correct.

[^4]: You might wonder why we can't just solve CNFSAT by transforming the formula
  into DNF and solving that. The answer is that the transformation can make the
  formula exponentially larger. So even though solving the problem on DNF is
  linear in the size the DNF formula, that size is *exponential* in the size of
  the CNF formula. So we still use exponential time in the size of the CNF formula.

[^5]: You might notice that this isn't *really* negation. After all, a type
  might have *neither* the method `X()` *nor* the method `X() int`. So, in the
  logic language we are building, we don't have
  [the law of the excluded middle](https://en.wikipedia.org/wiki/Law_of_excluded_middle).
  That's a problem we'll encounter again in a bit, but for this section it does
  not matter.

> [^10]: Yes, I know that there are complexity classes between polynomial and

Reviewer 2: Looks like this isn't referenced anywhere.

Author: Footnote within a footnote. Trying to be funny.

I would prefer it to appear directly after the first footnote, but hugo's markdown re-orders them in the order they appear in the text, not in the numerical order of the indices. So, I don't know how to do that. Better for the numbers and order to be consistent with the rendered output then.

---
date: "2022-05-12 08:00:00"
tags:
- golang
- programming
title: "Calculating type sets is harder than you think"
summary: I try to explain why the problem of calculating the exact contents of
  a type set in Go is harder than it might seem on the surface.
---

> Go 1.18 added the biggest and probably one of the most requested features of

Reviewer 6: Why “probaby”? The survey results quite clearly point out that it is indeed one of the most requested features, AFAIK :)

Author: Hedgers gonna hedge. If nothing else, it might subconsciously soothe people who kneejerk to any mention of generics.

all time to the language: [Generics](https://go.dev/doc/go1.18#generics). If
you want a comprehensive introduction to the topic, there are many out there
and I would personally [recommend this talk I gave at the Frankfurt Gopher
Meetup](https://www.youtube.com/watch?v=QP6v-Q5Foek).

This blog post is not an introduction to generics, though. It is about [this
sentence from the spec](https://go.dev/ref/spec#Operands):

> Implementation restriction: A compiler need not report an error if an
> operand's type is a type parameter with an empty type set.

This restriction might seem strange to you. After all, if a type set is empty,

> it seems very helpful to report that to the user, because they obviously made a

Reviewer: Please consider this and all my other remarks minor nits or even just stylistic preferences...
s/user, because they/user. They/

mistake - an empty type set can never be used as a constraint, as the function
using it could never be instantiated.

> I want to explain the reason that sentence is there and also go into a couple

Reviewer: s/the reason that sentence is there/why this sentence exists/

The reason that sentence is in the Go spec, is that it turns out to be hard to
actually determine if a type set is empty. Hard enough, that the Go team did
not want to require an implementation to solve that. Let's see why.

## P vs. NP

When we talk about whether or not a problem is hard, we often group problems
into two big classes:

1. Problems which can be *solved* reasonably efficiently. This class is called
   P.
2. Problems which can be *decided* reasonably efficiently. This class is called
   NP.

The first obvious follow up question is “what does ‘reasonably efficient’
mean?”. The answer to that is “there is an algorithm with a running time
polynomial in its input size”[^1].

The second obvious follow up question is “what's the difference between
‘solving’ and ‘deciding’?”.

*Solving* a problem means what you think it means: Finding a solution. If I
give you a number and ask you to solve the factorization problem, I'm asking
you to find a (non-trivial) factor of that number.

*Deciding* a problem means that I give you a solution and I'm asking you if the
solution is correct. For the factorization problem, I'd give you two numbers

> and ask you to verify that the second is a factor of the first.

Reviewer 5: This definition of "decided" is outside my experience.  Maybe instead of "decided", consider "Problems whose solutions can be verified reasonably efficiently."  (Though that leaves you at a loss for a nice single word description of that trait, so 🤷 .)  Or perhaps add it as an expansion in the "Deciding" paragraph.

Author: The issue is that "decision problem" is established jargon and I don't want to stray too far from that. Part of the intent of the post is to introduce this topic and this type of argument to people. If they read more about it, I want them to be able to connect it back to what they read here. So, this particular instance of jargon stays, I think.

Hopefully the additional context helps understand it.

These two things are often very different in difficulty. If I ask you to give
me a factor of 297863737, you probably know no better way than to sit down and
try to divide it by a lot of numbers and see if it comes out evenly. But if I
ask you to verify that 9883 is a factor of that number, you just have to do a
bit of long division and it either divides it, or it does not.

It turns out, that every problem which is efficiently *solvable* is also
efficiently *decidable*. You can just calculate the solution and compare it to
the given one, if you want to decide it. So, every problem in P is also in
NP[^2]. But it is [a famously open question](https://en.wikipedia.org/wiki/P_versus_NP_problem)
whether the opposite is true - that is, we don't really *know*, if there are
problems which are hard to solve, but easy to decide.

The reason this is hard to know in general, is that just because *we have not
found an efficient algorithm to solve a problem*, does not mean there is none.
In practice, though, we usually assume that there are some problems like that.

> One fact that helps us talk about hard problems, is that it turns out that

Reviewer: s/ it turns out that//

evaluating the formula. For example

    X = true
    Y = false
    Z = true
    ((¬X∧Y)∨Z)∧(X∨¬Y) = ((¬true ∧ false) ∨ true)∧(true ∨ ¬false)
                      = ((false ∧ false) ∨ true)∧(true ∨ true)
                      = (false ∨ true) ∧ true
                      = true ∧ true
                      = true

But if I *only* give you the formula and ask you to *find* an assignment which
makes it `true` - or even to find whether such an assignment exists - you
probably have to try out all possible assignments to see if any of them does.
That's easy for three variables, but there are 2<sup>n</sup> possible
assignments, so it takes *exponential* time in the number of variables.

This problem is called "boolean satisfiability" and it is NP-complete.

It turns out to be extremely important *in what form* the formula is given,
though. Some forms make it pretty easy to solve, while others make it hard.

For example, every formula can be rewritten into what is called a [“Disjunctive Normal Form” (DNF)](https://en.wikipedia.org/wiki/Disjunctive_normal_form),
so called because it consists of a series of conjunction (“and”) terms, joined

> together by disjunction (“or”) operators[^3]:

Reviewer 4: Could you add some indication which is `and` and which is `or` for dummies like me?
```suggestion
so called because it consists of a series of conjunction (“and”, `∧`) terms, joined
together by disjunction (“or”, `∨`) operators[^3]:
```

Reviewer: Maybe even a table with the variants - (conjunction and ∧ &&), (disjunction or ∨ ||), (negation not ¬ !)

    S[f] ∪ S[g] = { x∈U | f(x) } ∪ { x∈U | g(x) }
                = { x∈U | f(x) ∨ g(x) }

    f[S](x) ∨ f[T](x) = x∈S ∨ x∈T = x∈(S∪T)

Similarly, `∩` becomes equivalent to `∧`. `¬` becomes equivalent to set
complement: `S[¬f] = { x∈U | ¬f(x) } = { x∈U | x∉S }`, which we can write as
`U\S`.

The boolean formulas we looked at in the previous section are a form of these
boolean functions. They take some boolean arguments and return a boolean, e.g.

      f: B³ → B
    (X,Y,Z) ↦ ((¬X∧Y)∨Z)∧(X∨¬Y)

[It turns out](https://en.wikipedia.org/wiki/Functional_completeness) that *every*
boolean function with boolean parameters can be written as a logical formula -
in particular, every such function has a CNF and a DNF.

If we want to do that for other sets (apart from `B^n`, that is), we first have
to be able to translate it into a set of boolean variables. As we will see,
this can be easily done in the case we are interested in.

All of this allows us to talk about what we mean by “computing a set is hard”.
With this equivalency, checking if a given object is in a set becomes the
*decision* problem of SAT - deciding if `x∈S` is equivalent to deciding if

> `f[S](x)`evaluates to `true`. And checking if a set is empty becomes the

Reviewer 4: ```suggestion
`f[S](x)` evaluates to `true`. And checking if a set is empty becomes the
```

*solution* problem of SAT, as it is equivalent to verifying that `f[S]` is not
satisfiable.

With this, let us look at the specific sets we are interested in.

## Basic interfaces as type sets

Interfaces in Go have always described sets of types. For example, the interface

```go
type S interface {
    X()
    Y()
    Z()
}
```

is “the set of all types which have a method `X()` and a method `Y()` and a method `Z()`”.

We also have a way to express set intersection, using [interface embedding](https://go.dev/ref/spec#Embedded_interfaces):

```go
type S interface { X() }
type T interface { Y() }
type U interface {
    S
    T
}
```

> This expresses `T=S∩T` as an interface. We can also view the predicate “has a

Reviewer 5: Do you mean `U=S∩T`?

Author: Yes, thanks

method `X()`” as a boolean variable and think of this as the formula `X∧Y`.

Surprisingly, there is also a simple form of negation. It happens implicitly,
because a type can not have two methods with the same name, but different
types. So, implicitly, if a type has a method `X()`, it does *not* have a
method `X() int`, for example[^5]:

```go
type X interface { X() }

> type NotX interface{ X() int }

Reviewer 5: There are, of course, infinitely many other interfaces that could also be `NotX`.

Author: Do you think this needs another mention? In my defense, I do say "for example".

Reviewer 5: No, I suppose not.

```

This meant that we can express interfaces which could never be satisfied by
any type (i.e. which describe an empty type set):

```go
interface{ X; NotX }
```

[The compiler rejects such interfaces](https://go.dev/play/p/r4kpXNynscX).

So we had a language to build boolean formulas, where every formula had the
form

    X∧Y∧¬Z

Checking if a type implements an interface means checking if it is in the set
described by that interfaces. Which means checking if the corresponding formula
is satisfied.

And checking if an interface is invalid meant checking if the set it describes
is empty. Which meant checking if the formula is *satisfiable*. This was easy,
though, as those formulas are in DNF. They only have a single term and that
term is a conjunction of variables or their negations. And as we noted above,

> solving SAT on a formula given in DNF is easy.

Reviewer 5: These three paragraphs switch between present & past tense pretty fluidly and it's confusing.

```
So we HAVE a language to build boolean formulas, where every formula HAS the
form

    X∧Y∧¬Z

Checking if a type implements an interface means checking if it is in the set
described by that interfaces. Which means checking if the corresponding formula
is satisfied.

And checking if an interface is invalid meanS checking if the set it describes
is empty. Which meanS checking if the formula is *satisfiable*. This IS easy,
though, as those formulas are in DNF. They only have a single term and that
term is a conjunction of variables or their negations. And as we noted above,
solving SAT on a formula given in DNF is easy.
```

Author: I blame Roger.
Fixed it.

> embed comparable or interfaces that specify methods.

This disallows the main mechanism we used to map formulas to interfaces above.
We can no longer express our `TertiumNonDatur` type, or the individual `|`
terms of the formula, as the respective terms specify methods. Without
specifying methods, we can't get our “implicit negation” to work either.

There are a couple of other restrictions which are arguably at least touching
on these problem, but I won't got into them here.

Ultimately, the fix was to restrict the language so that hard to manage
interfaces could no longer be expressed to begin with.

## What now?

Maybe you noticed a bit of a bait-and-switch I pulled. Originally, I said that
I'm going to explain why the Go compiler does not complain about empty type
sets. I claimed that checking that is NP-complete. But in the last section, I
told you that the NP-completeness problem was fixed. Could the compiler
complain after all?

To the best of my knowledge, the answer is “we don't know yet”. It is entirely
possible that with these restrictions, we would efficiently implement that
check. My intuition is, that it should be possible. But I am not aware that

> anyone sat down and either

Reviewer 5: I mean, maybe they remained standing.  ;)

1. Wrote an algorithm to prove that a type set is empty, thus proving that it
   can be done efficiently. Or
2. Wrote a proof that the problem is still NP-complete.

The issue is, that doing either of these takes time. I did invest that time in

> the original discussion, before the limitations where introduced and so far, I

Reviewer 6: where → were

  polynomial. And for many practical problems, even n<sup>3</sup> is
  intolerably slow. But for complicated reasons, there is a qualitatively
  important difference between “polynomial” and “exponential”[^6] run time, so
  just trust me, that the distinction makes sense.

[^2]: These names might seem strange, by the way. P is easy to explain: It
  stands for “polynomial”. NP doesn't mean “not polynomial”, though. It means
  “non-deterministic polynomial”. A non-deterministic computer, in this
  context, is a hypothetical machine which can run arbitrarily many
  computations simultaneously. A program which can be *decided* efficiently by
  any computer, can be *solved* efficiently by a non-deterministic one, by just
  trying out all possible solutions at the same time and returning a correct
  one. Thus, being able to decide a problem on a normal computer, means being
  able to solve it on a non-deterministic one.

[^3]: You might complain that it is hard to remember if the “disjunctive normal
  form” is a disjunction of conjunctions, or a conjunction of disjunctions -
  and that no one can remember which of these means “and” and which means “or”

>   anyways. You would be correct.

Reviewer 5: 😆

Author: I honestly have to look this up every. Single. Time.
So I'm joking, but I'm also not. This is absolute garbage jargon.
Unfortunately, I'm generally stuck with using it, I think.

Reviewer 5: I haven't had to deal with these terms very often, but it didn't seem that bad to me?  "X normal form" seems clear that it means "a list of things joined by X": "Conjunctive Normal Form" => "a list of things joined by conjunctions". 
 `Join([]interface{}, &&)`, so to speak 😆

---
date: "2022-05-12 08:00:00"
tags:
- golang
- programming
title: "Calculating type sets is harder than you think"
summary: I try to explain why the problem of calculating the exact contents of
  a type set in Go is harder than it might seem on the surface.
---

<script defer crossorigin="anonymous" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.2/MathJax.js?config=TeX-MML-AM_CHTML"></script>

Go 1.18 added the biggest and probably one of the most requested features of
all time to the language: [Generics](https://go.dev/doc/go1.18#generics). If
you want a comprehensive introduction to the topic, there are many out there
and I would personally [recommend this talk I gave at the Frankfurt Gopher
Meetup](https://www.youtube.com/watch?v=QP6v-Q5Foek).

This blog post is not an introduction to generics, though. It is about [this
sentence from the spec](https://go.dev/ref/spec#Operands):

> > Implementation restriction: A compiler need not report an error if an

Reviewer 7: Passing comment: it seems odd to call this a "restriction", given that it sounds more like an option to relax?

Author: Yeah, I also find the phrasing a bit weird. "A compiler need not", to me, sounds like "a compiler does not have to". When really, it should probably be "a compiler should not" (probably as a "it is valid for an interface to have an empty type set" in the type set section).

I'm not sure what the reasoning here is. But I think I've been vocal enough about my opinion in the discussion and I don't want to nitpick.

> operand's type is a type parameter with an empty type set.

> 

Reviewer 7: At this point I would be tempted to give people a quick refresher on what a type set is. 

"As a quick refresher..." 

with a code sample.

Author: I'm adding an example of an interface with an empty type set, which hopefully is enough of a reminder, while also not taking up too much time. It's probably helpful anyways, to demonstrate the problem.

This restriction might seem strange to you. After all, if a type set is empty,
it seems very helpful to report that to the user, because they obviously made a
mistake - an empty type set can never be used as a constraint, as the function

> using it could never be instantiated.

Reviewer 7: With the quick refresher and example on type sets (which will help make a more gentle connection from any introduction to generics), you might consider rephrasing this paragraph as a sequence of questions. Perhaps with the gist of "But hang on, given the example above, wouldn't it be important for a user to know when a type set is empty? If we don't report an error in such a situation, doesn't that make DX a bit ugly?"

Author: It would probably read better but while trying to do it it felt unnatural to me. For no particular reason.
At this point the marginal benefits of further editing start to diminish anyways. So I think I will leave it.

> I want to explain the reason that sentence is there and also go into a couple

Reviewer 7: Personal preference, but I would start this sentence with "This article..." instead of "I". It makes clear to the reader that you are just about to explain the raison d'être of this article. If you also phrase the previous paragraph in the form of questions, this paragraph can be seen as (part of) the answer, or at least laying the ground for the answer.

Author: I think I prefer to stick with I here. I want to prevent sounding too clinical.

of related design decisions of the generics design. I'm trying to be expansive
in my explanation, which means that you should not *need* any special knowledge
to understand it. It also means, some of the information might be boring to
you - feel free to skip the corresponding sections.

> The reason that sentence is in the Go spec, is that it turns out to be hard to

Reviewer 7: If you did follow the "question" and "answer" proposal above, this paragraph, minus the last sentence, goes after the answer, before the paragraph "This article..."

       - If it contains `!x`, then we must pass `false` for `x`
       - If it contains neither, then what we pass for `x` does not matter and
         either value works.
   * The term then evaluates to `true` with these arguments, so the entire
     expression does.
2. If none of the terms can be made `true`, the function can never return
   `true` and there is no valid set of arguments.

On the other hand, there is also a [“Conjunctive Normal Form”
(CNF)](https://en.wikipedia.org/wiki/Conjunctive_normal_form). Here, the
expression is a series of *disjunction* (`||`) terms, joined together with
*conjunction* (`&&`) operators:

```go
func F_CNF(x, y, z bool) bool {
  return (!x || z) && (y || z) && (x || !y)
}
```

(Again, you can [verify that this is the same function](https://go.dev/play/p/0xldLVGqu7m))

For this, the idea of our algorithm does not work. To find a solution, you have
to take *all terms* into account simultaneously. You can't just tackle them one
by one. In fact, solving satisfiability on CNF (often abbreviated as “CNFSAT”)
is NP-complete[^4].

[It turns out](https://en.wikipedia.org/wiki/Functional_completeness) that

> *every* boolean function can be written as a single expression using only `||`, `&&` and `!`. In particular, every boolean has a DNF and a CNF.

Reviewer 6: This should be “every boolean function”, right? The word “function” is missing.

Author: indeed.

implicitly, if a type has a method `X()` it does *not* have a method `X()
int` for example[^6]:

```go
type X interface { X() }
type NotX interface{ X() int }
```

There is a small snag: A type can have *neither* a method `X()` *nor* have a
method `X() int`. That's why our negation operator is limited. Real boolean
variables are always *either* `true` *or* `false`, whereas our negation also
allows them to be neither. In mathematics we say that this logic language lacks
[the law of the excluded middle](https://en.wikipedia.org/wiki/Law_of_excluded_middle)
(also called “Tertium Non Datur” - “there is no third”). For this section, that does not matter. But we have to worry about it later.

Because we have intersection and negation, we can express interfaces which
could never be satisfied by any type (i.e. which describe an empty type set):

```go
interface{ X; NotX }
```

[The compiler rejects such interfaces](https://go.dev/play/p/r4kpXNynscX). But
how can it do that? Did we not say above that checking if a set is empty is
NP-complete?

The reason this works is that we only have negation and conjunction (`&&`). So,

> all the boolean expressions e can build with this language have the form

Reviewer 7: Did you mean `e`?

Author: No, I meant "we" ^^

The hope is that this change (among a couple of others) is sufficient to ensure
that we can always calculate type sets accurately. Which means I pulled a bit
of a bait-and-switch: I said that calculating type sets is hard. But as they
actually exist, they *might not be*.

The reason I wrote this blog post anyways is to explain the *kind of problems*
that exist in this area. It is easy to say we have solved this problem
[once and for all](https://www.youtube.com/watch?v=0SYpUSjSgFg).

But to be certain, someone should *prove* this - either by writing a proof like
this post, that the problem is still hard, or by writing an algorithm which
solves it efficiently.

There are also still discussions about changing the generics design. As one
example, the limitations we introduced to fix all of this made
[one of the use cases from the design doc](https://go.googlesource.com/proposal/+/refs/heads/master/design/43651-type-parameters.md#interface-types-in-union-elements)
impossible to express. We might want to tweak the design to allow this use
case. In these discussions, we have to look out for this, so we don't
re-introduce the problem. It took us some time to even detect it

> [when the union operator was proposed](https://github.com/golang/go/issues/45346)a.

Reviewer 7: Trailing 'a' after link

And there are other kinds of “implicit negations” in the Go language. For
example, a `struct` can not have both a field *and* a method with the same
name. Or being one type implies not being another type (so `interface{int}`
implicitly negates `interface{string}`).

All of which is to say that even if the problem might no longer be
NP-complete - I hope that I convinced you it is still more complicated than you
might have thought.

---

I want to thank my beta-readers for helping me improve this article. In
particular,
[arnehormann](https://github.com/arnehormann),
[@johanbrandhorst](https://twitter.com/johanbrandhorst),
[@mvdan_](https://twitter.com/mvdan_),
[@readcodesing](https://mobile.twitter.com/readcodesing),
[@rogpeppe](https://twitter.com/rogpeppe) and
[@zekjur](https://twitter.com/zekjur).

> They took a significant chunk of time and their suggestions where invaluable.

Reviewer 7: s/where/were/

Reviewer: also, missing @myitcv :-)
